fusillade/api/evaluate.py

```diff
@@ -1,28 +1,58 @@
-from dcplib.aws import clients as aws_clients
-from flask import make_response, request, jsonify
+from threading import Thread
+from concurrent.futures import Future
+import typing
+
+from flask import make_response, jsonify
 
 from fusillade import User, directory
 
-iam = aws_clients.iam
-
-
-def evaluate_policy():
-    json_body = request.json
-    principal = json_body["principal"]
-    action = json_body["action"]
-    resource = json_body["resource"]
-    user = User(directory, principal)
-    result = iam.simulate_custom_policy(
-        PolicyInputList=user.lookup_policies(),
-        ActionNames=[action],
-        ResourceArns=[resource],
-        ContextEntries=[
-            {
-                'ContextKeyName': 'fus:user_email',
-                'ContextKeyValues': [principal],
-                'ContextKeyType': 'string'
-            }
-        ]
-    )['EvaluationResults'][0]['EvalDecision']
-    result = True if result == 'allowed' else False
-    return make_response(jsonify(principal=principal, action=action, resource=resource, result=result), 200)
+from fusillade.utils.authorize import assert_authorized, evaluate_policy
+
+
+def evaluate_policy_api(user, body):
+    with ThreadedAuthorize(user, ['fus:Evaluate'], ['arn:hca:fus:*:*:user']):
+        policies = User(directory, body['principal']).lookup_policies()
+        result = evaluate_policy(body['principal'], body['action'], body['resource'], policies)
+    return make_response(jsonify(**body, result=result), 200)
+
+
+def call_with_future(fn, future, args, kwargs):
+    try:
+        result = fn(*args, **kwargs)
+        future.set_result(result)
+    except Exception as exc:
+        future.set_exception(exc)
+
+
+def threaded(fn):
+    def wrapper(*args, **kwargs):
+        future = Future()
+        Thread(target=call_with_future, args=(fn, future, args, kwargs)).start()
+        return future
+
+    return wrapper
+
+
+class ThreadedAuthorize:
+    """
+    Allows us to authorize the request and evaluate the request in parallel.
+    """
+    def __init__(
+            self,
+            user: str,
+            actions: typing.List[str],
+            resources: typing.List[str]
+    ):
+        self.user = user
+        self.actions = actions
+        self.resources = resources
+
+    def __enter__(self):
+        self.future = self._evaluate()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.future.result()
+
+    @threaded
+    def _evaluate(self):
+        assert_authorized(self.user, self.actions, self.resources)
```

Comment:
Can we make threaded a static method of this class? Would be easier to follow the flow of control